# **Car Price by Feature Analysis**

## Objectives

* The purpose of this project is to run a market pricing analysis of cars on the American market, based on a number of features, from brand, model, risk, engineering features and associated costs. The independent variables are analysed in order to find relationships between them and determine those features which predict the price most reliably. The analysis is explained thoroughly such that multiple business, pricing analyst, marketing analyst and engineer audiences can take away industry relevant messages.

## Inputs

* The dataset can be found on [Kaggle](https://www.kaggle.com/datasets/hellbuoy/car-price-prediction/data). The dataset was downloaded on a personal machine in the same folder as this jupyter notebook, and uploaded to GitHub (see CarPrice_Assignment.csv).

## Outputs

* The notebook will not generate additional files but at least one new dataset with cleaned, removed and added features might be produced, along with associated visualisations and models.

## Additional Comments

* This notebook runs on the base Python environment of 3.11.7, since the virtual Python environment does not apply to the parent folder of this jupyter notebook. The necessary Python libraries were not installed again for the sake of this notebook (in a separate virtual environment or to match the required version of Python, 3.12.8, because of potential Python version clashes).



---

# Change working directory

* We are assuming you will store the notebooks in a subfolder, therefore when running the notebook in the editor, you will need to change the working directory
(Comment: this disabled the root directory's venv)

We need to change the working directory from its current folder to its parent folder
* We access the current directory with os.getcwd()

In [1]:
import os
current_dir = os.getcwd()
current_dir

'/Users/mariaoanacucu/Documents/VS Code Projects/Car_feature_price_analysis/jupyter_notebooks'

We want to make the parent of the current directory the new current directory
* os.path.dirname() gets the parent directory
* os.chir() defines the new current directory

In [2]:
os.chdir(os.path.dirname(current_dir))
print("You set a new current directory")

You set a new current directory


Confirm the new current directory

In [3]:
current_dir = os.getcwd()
current_dir

'/Users/mariaoanacucu/Documents/VS Code Projects/Car_feature_price_analysis'

---

# Load the data

Import Python libraries

In [4]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

In [6]:
df = pd.read_csv('../CarPrice_Assignment.csv')

---

# Inspect the data

Check data format, shape, types, columns and missing values

First, it is useful to check the entire structure of the dataset, as is, by calling df

In [3]:
df

,car_ID,symboling,CarName,fueltype,aspiration,doornumber,carbody,drivewheel,enginelocation,wheelbase,...,enginesize,fuelsystem,boreratio,stroke,compressionratio,horsepower,peakrpm,citympg,highwaympg,price
0,1,3,alfa-romero giulia,gas,std,two,convertible,rwd,front,88.6,...,130,mpfi,3.47,2.68,9.0,111,5000,21,27,13495.0
1,2,3,alfa-romero stelvio,gas,std,two,convertible,rwd,front,88.6,...,130,mpfi,3.47,2.68,9.0,111,5000,21,27,16500.0
2,3,1,alfa-romero Quadrifoglio,gas,std,two,hatchback,rwd,front,94.5,...,152,mpfi,2.68,3.47,9.0,154,5000,19,26,16500.0
3,4,2,audi 100 ls,gas,std,four,sedan,fwd,front,99.8,...,109,mpfi,3.19,3.40,10.0,102,5500,24,30,13950.0
4,5,2,audi 100ls,gas,std,four,sedan,4wd,front,99.4,...,136,mpfi,3.19,3.40,8.0,115,5500,18,22,17450.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
200,201,-1,volvo 145e (sw),gas,std,four,sedan,rwd,front,109.1,...,141,mpfi,3.78,3.15,9.5,114,5400,23,28,16845.0
201,202,-1,volvo 144ea,gas,turbo,four,sedan,rwd,front,109.1,...,141,mpfi,3.78,3.15,8.7,160,5300,19,25,19045.0
202,203,-1,volvo 244dl,gas,std,four,sedan,rwd,front,109.1,...,173,mpfi,3.58,2.87,8.8,134,5500,18,23,21485.0
203,204,-1,volvo 246,diesel,turbo,four,sedan,rwd,front,109.1,...,145,idi,3.01,3.40,23.0,106,4800,26,27,22470.0


Calling df.info() will give a summary of the data, including numbers of rows, columns, NaN entries and data types.

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 205 entries, 0 to 204
Data columns (total 26 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   car_ID            205 non-null    int64  
 1   symboling         205 non-null    int64  
 2   CarName           205 non-null    object 
 3   fueltype          205 non-null    object 
 4   aspiration        205 non-null    object 
 5   doornumber        205 non-null    object 
 6   carbody           205 non-null    object 
 7   drivewheel        205 non-null    object 
 8   enginelocation    205 non-null    object 
 9   wheelbase         205 non-null    float64
 10  carlength         205 non-null    float64
 11  carwidth          205 non-null    float64
 12  carheight         205 non-null    float64
 13  curbweight        205 non-null    int64  
 14  enginetype        205 non-null    object 
 15  cylindernumber    205 non-null    object 
 16  enginesize        205 non-null    int64  
 1

We will break down the above information by investigating the shape, column features, as well as missing and repeated values separately.

In [8]:
df.shape

(205, 26)

df.shape indicates the tabular structure of the data set, with 205 rows and 26 columns.

In [9]:
df.columns.to_list()

['car_ID',
 'symboling',
 'CarName',
 'fueltype',
 'aspiration',
 'doornumber',
 'carbody',
 'drivewheel',
 'enginelocation',
 'wheelbase',
 'carlength',
 'carwidth',
 'carheight',
 'curbweight',
 'enginetype',
 'cylindernumber',
 'enginesize',
 'fuelsystem',
 'boreratio',
 'stroke',
 'compressionratio',
 'horsepower',
 'peakrpm',
 'citympg',
 'highwaympg',
 'price']

Let's have a look at the features provided in the columns and their significance. It is useful to do this at the beginning of the ETL process in order to understand the data we are working with. Applying a semantic discriminant analysis with the help of ChatGPT (asking it to group variables based solely on column names), and then manually after review, we can easily group the features into the following: 
1. Identifiers and labels (car_ID, CarName, price)
2. Insurance/Risk Information (symboling)
3. Car body configuration (doornumber, carbody, drivewheel, enginelocation)
4. Car dimensions and physical size (wheelbase, carlength, carwidth, carheight, curbweight)
5. Engine design (fueltype, aspiration, enginetype, cylinder number, enginesize, fuelsystem)
6. Internal engine measurements (boreratio, stroke, compressionratio)
7. Engine performance (horsepower, peakrpm)
8. Fuel efficiency (citympg, highwaympg)
9. The outcome variable: price


This could be repeated compared at the end with an unsupervised mixed factor analysis, to reveal if feature category is indicative or not of the price. However, this sort of complex analysis is beyond the scope of the current assignment, so we use it only for the purpose of data comprehension of a general business audience.

In [10]:
df.isnull().sum()

car_ID              0
symboling           0
CarName             0
fueltype            0
aspiration          0
doornumber          0
carbody             0
drivewheel          0
enginelocation      0
wheelbase           0
carlength           0
carwidth            0
carheight           0
curbweight          0
enginetype          0
cylindernumber      0
enginesize          0
fuelsystem          0
boreratio           0
stroke              0
compressionratio    0
horsepower          0
peakrpm             0
citympg             0
highwaympg          0
price               0
dtype: int64

In [11]:
df.duplicated().sum()

0

The dataset contains no missing or duplicated values, as shown by counting the isnull() and duplicated() methods of df. 

Let's further inspect the data by looking at unique values and their numbers, in order to fully understand it.

In [12]:
#print number of unique values for each column
for col in df.columns:
    print(col, '--', df[col].nunique())

car_ID -- 205
symboling -- 6
CarName -- 147
fueltype -- 2
aspiration -- 2
doornumber -- 2
carbody -- 5
drivewheel -- 3
enginelocation -- 2
wheelbase -- 53
carlength -- 75
carwidth -- 44
carheight -- 49
curbweight -- 171
enginetype -- 7
cylindernumber -- 7
enginesize -- 44
fuelsystem -- 8
boreratio -- 38
stroke -- 37
compressionratio -- 32
horsepower -- 59
peakrpm -- 23
citympg -- 29
highwaympg -- 30
price -- 189


In [13]:
#print all unique values of each column
for col in df:
    print(col, '--', df[col].unique())

car_ID -- [  1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16  17  18
  19  20  21  22  23  24  25  26  27  28  29  30  31  32  33  34  35  36
  37  38  39  40  41  42  43  44  45  46  47  48  49  50  51  52  53  54
  55  56  57  58  59  60  61  62  63  64  65  66  67  68  69  70  71  72
  73  74  75  76  77  78  79  80  81  82  83  84  85  86  87  88  89  90
  91  92  93  94  95  96  97  98  99 100 101 102 103 104 105 106 107 108
 109 110 111 112 113 114 115 116 117 118 119 120 121 122 123 124 125 126
 127 128 129 130 131 132 133 134 135 136 137 138 139 140 141 142 143 144
 145 146 147 148 149 150 151 152 153 154 155 156 157 158 159 160 161 162
 163 164 165 166 167 168 169 170 171 172 173 174 175 176 177 178 179 180
 181 182 183 184 185 186 187 188 189 190 191 192 193 194 195 196 197 198
 199 200 201 202 203 204 205]
symboling -- [ 3  1  2  0 -1 -2]
CarName -- ['alfa-romero giulia' 'alfa-romero stelvio' 'alfa-romero Quadrifoglio'
 'audi 100 ls' 'audi 100ls' 'audi fox' 'au

---

# Select and group features

Let's proceed by understand data types so as to group or replace them with the appropriate values.

In [14]:
df.dtypes

car_ID                int64
symboling             int64
CarName              object
fueltype             object
aspiration           object
doornumber           object
carbody              object
drivewheel           object
enginelocation       object
wheelbase           float64
carlength           float64
carwidth            float64
carheight           float64
curbweight            int64
enginetype           object
cylindernumber       object
enginesize            int64
fuelsystem           object
boreratio           float64
stroke              float64
compressionratio    float64
horsepower            int64
peakrpm               int64
citympg               int64
highwaympg            int64
price               float64
dtype: object

---

# Conclusions and next steps

* In cases where you don't need to push files to Repo, you may replace this section with "Conclusions and Next Steps" and state your conclusions and next steps.

In [ ]:
import os
try:
  # create your folder here
  # os.makedirs(name='')
except Exception as e:
  print(e)
